### PR TITLE
[InstCombine] Change (add x, c) to (xor x, c)

### PR DESCRIPTION
**Pull request by Peter9606**
Change (add x, c) to (xor x, c) iff c is constant and c equals the top bit of the demanded bits.
Alive2: https://alive2.llvm.org/ce/z/DKmkwF

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3351]: It also holds for vectors.
```suggestion
  
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3356]:
```suggestion
              m_c_Or(m_And(m_Add(m_Value(X), m_APInt(HalfC1)), m_APInt(C3)),
                     m_And(m_Value(X), m_APInt(C2))))) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3358]:
```suggestion
      KnownBits KnownX = computeKnownBits(X, 0, &I);
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3362]:
```suggestion
        return BinaryOperator::CreateXor(X, ConstantInt::get(Ty, *HalfC1));
```

File: llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll:18]: Please add some negative tests.

File: llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll:27]:
```suggestion
  %add = add i16 %x, 32
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
This pattern looks complex. Does it come from a real-world application?

**Comment by Peter9606** [comment]
> This pattern looks complex. Does it come from a real-world application?

Yes, it does.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (Peter9606) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3362]: Thanks for your quick review and comments. Not quite following this piece comment, could you please explain a little bit more on what is the difference betwee those two forms? Thansk in advance!

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (Peter9606) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: fixed!

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (Peter9606) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: done!

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (Peter9606) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: fixed!

File: llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll
Comment (Peter9606) [reply llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll]: done!

File: llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll
Comment (Peter9606) [reply llvm/test/Transforms/InstCombine/or-and-add-constant-constant-and-constant.ll]: done!

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (bcl5980) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3357]: Check constant first, computeKnownBits is a heavy function call.

**Comment by bcl5980** [comment]
Precommit test First, then second commit is code change + test diff.
It looks your test is the second commit.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (bcl5980) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3343]: maybe move into foldComplexAndOrPatterns?

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (Peter9606) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Done!

**Comment by Peter9606** [comment]
> Precommit test First, then second commit is code change + test diff. It looks your test is the second commit.

No sure how it's shown in github PR, but I checked in my local environment that the order is correct, precommit first, then the patch.

**Comment by Peter9606** [comment]
> Precommit test First, then second commit is code change + test diff. It looks your test is the second commit.

You are right, fixed! Thanks!

**Comment by bcl5980** [comment]
We may need to rewrite the conditions.
https://alive2.llvm.org/ce/z/-ct9Wt

Not all cases can be catched for now.

**Comment by topperc** [comment]
I think (and (add x, 32), 48) can be replaced with (and (xor x, 32), 48). Is that optimization alone enough to trigger the rest?

**Comment by Peter9606** [comment]
> I think (and (add x, 32), 48) can be replaced with (and (xor x, 32), 48). Is that optimization alone enough to trigger the rest?

Just did a quick verify, your sugguestion works. At least for the case pointed by @bcl5980  in [comment](https://github.com/llvm/llvm-project/pull/75129#issuecomment-1852209646), it can be simplified to XOR now.

**Comment by Peter9606** [comment]
> I think (and (add x, 32), 48) can be replaced with (and (xor x, 32), 48). Is that optimization alone enough to trigger the rest?

@topperc  If I understand correctlly, your suggestion can be formalized as https://alive2.llvm.org/ce/z/aoJPsn, right?

**Comment by nikic** [comment]
@Peter9606 We'd probably approach that as a demanded bits transform (https://github.com/llvm/llvm-project/blob/d293a354d076c772ccd2972a71cb4a7d6f9aa410/llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp#L492). Convert `X + C` to `X ^ C` if `C` is equal to the top bit of the `DemandedMask`.

**Comment by Peter9606** [comment]
> @Peter9606 We'd probably approach that as a demanded bits transform (
> 
> https://github.com/llvm/llvm-project/blob/d293a354d076c772ccd2972a71cb4a7d6f9aa410/llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp#L492
> 
> ). Convert `X + C` to `X ^ C` if `C` is equal to the top bit of the `DemandedMask`.

Thanks for your detail guidance. @nikic  
PR is updated!

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp
Comment (topperc) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:560]: You can use `DemandedMask.getActiveBits() - 1`

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:559]:
```suggestion
      if (match(I->getOperand(1), m_APInt(C)) &&
```
We already know that this is an `add`, so I think we can just check the operand here.

File: llvm/test/Transforms/InstCombine/and.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/and.ll:2746]: Can you please also add a test for the larger motivating pattern you originally wanted to handle? Just to show that it does get handled as part of this as well...

**Review by nikic — COMMENTED** [review]
Looks reasonable to me.

**Comment by bcl5980** [comment]
Add a vector case?

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp
Comment (Peter9606) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: updated! thank you!

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp
Comment (Peter9606) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: done!

File: llvm/test/Transforms/InstCombine/and.ll
Comment (Peter9606) [reply llvm/test/Transforms/InstCombine/and.ll:2757]: original case appended.

**Review by nikic — APPROVED** [review]
LGTM

**Review by dtcxzyw — APPROVED** [review]
LGTM. I will merge this PR after the CI is complete.

**Comment by jplehr** [comment]
I see errors in our buildbot (https://lab.llvm.org/buildbot/#/builders/193/builds/43753) after this patch landed.

Reverting this patch locally resolves the errors.
If you need additional info / help for reproducing, please let me know.

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:560]:
```
Instruction *Xor = BinaryOperator::CreateXor(I->getOperand(0), ConstantInt::get(VTy, *C));
return InsertNewInstWith(Xor, I->getIterator());
```
I guess it will fix the assertion failure.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:560]: Ah, sorry for missing that. As this is in SimplifyDemanded we should explicitly set the IRBuilder InsertPoint.

**Comment by nikic** [comment]
Should be fixed by https://github.com/llvm/llvm-project/commit/9100228e455bfe992e291e060c5b697d8403956e.

**Pull request merged**
